ykbh/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+#![feature(test)]
+extern crate test;
+
+use ykpack::{self, Local, Terminator, Statement, IPlace, Constant, ConstantInt, UnsignedInt};
+use yktrace::sir::SIR;
+use std::convert::TryInto;
+
+const ZST: Object = Object { data: Vec::new() };
+
+#[derive(Clone, Debug)]
+/// A data object, storing bytes.
+pub struct Object {
+    data: Vec<u8>
+}
+
+impl Object {
+
+    /// Creates a new Object from a `usize`.
+    fn from_usize(ptr: usize) -> Object {
+        let data = ptr.to_be_bytes();
+        Object {
+            data: Vec::from(data)
+        }
+    }
+
+    /// Creates a new Object from a `u8`.
+    fn from_u8(val: u8) -> Object {
+        Object {
+            data: Vec::from([val])
+        }
+    }
+
+    /// Converts the objects data back to a `usize`.
+    fn to_usize(&self) -> usize {
+        usize::from_be_bytes(self.data.as_slice().try_into().expect("Unexpected data length."))
+    }
+
+    /// Given a pointer and size, creates a new object by copying `size` data from `ptr`.
+    fn from_ptr(ptr: usize, size: usize) -> Object {
+        let mut data = Vec::with_capacity(size);
+        unsafe {
+            std::ptr::copy(ptr as *const u8, data.as_mut_ptr(), size);
+            data.set_len(size);
+        }
+        Object {
+            data
+        }
+    }
+}
+
+pub struct SIRInterpreter {
+    vars: Vec<Option<Object>>,
+}
+
+impl SIRInterpreter {
+    pub fn new(vars: Vec<Option<Object>>) -> Self {
+        SIRInterpreter {
+            vars,
+        }
+    }
+
+    pub fn interpret(&mut self, body: &ykpack::Body) {
+        // Ignore yktrace::trace_debug.
+        if body.flags & ykpack::bodyflags::TRACE_DEBUG != 0 {
+            return;
+        }
+
+        for stmt in body.blocks[0].stmts.iter() {
```

Comment:
So we only interpret one hard-coded block for now (hence there is no program counter)?

Reply:
Yes, this only interprets one block at the moment. I imagined that later when we do a call, we store the current frame, reinitialise the interpreter, and run it on the other block. Upon returning we just load back the old frame. Though I'm not sure what we'd need the program counter for?

Reply:
Well, you kind of have a program counter here. This for loop implicitly is one.

I expected the interpreter to have a stack of frames, then we make a call, we push a new frame (another collection of variables) and update the program counter to the first SIR block of the corresponding function.

So whereas this preliminary version of the interpreter deals with only one-frame's worth of variables, I think the interpreter should maintain a stack. Then we only ever need one instance of the interpreter.

Reply:
Ah ok, I see what you mean now. We need the counter to initialise the interpreter when a guard fails so we can start in the middle of a frame, and also when we come back from a call to remember where we left off.

Reply:
Ah yes. That too!